inc/azure_macro_utils/macro_utils.h

```diff
@@ -202,6 +202,61 @@ const char* MU_C3(MU_, enumIdentifier,_ToString)(enumIdentifier value)
 
 #define MU_ENUM_VALUE_2(enumIdentifier, value) "", MU_ENUM_TO_STRING_2(enumIdentifier, (value)), (int)(value)
 
+#define CONVERT_ENUM_VALUE(from, to) \
+    case (from): \
+        *enum_value_to = (to); \
+        return 0; \
+
+// This macro declares an enum conversion function that translates a value from one enum to another
+#define MU_DECLARE_CONVERT_ENUM(ENUM_TYPE_FROM, ENUM_TYPE_TO) \
+    int MU_C4(convert_enum_, ENUM_TYPE_FROM, _, ENUM_TYPE_TO)(ENUM_TYPE_FROM enum_value_from, ENUM_TYPE_TO* enum_value_to);
+
+// This macro defines an enum conversion function that translates a value from one enum to another
+// It is implemented as a switch statement. This allows for catching multiple from values being in the list also
+#define MU_DEFINE_CONVERT_ENUM_WITHOUT_INVALID(ENUM_TYPE_FROM, ENUM_TYPE_TO, ...) \
+    int MU_C4(convert_enum_, ENUM_TYPE_FROM, _, ENUM_TYPE_TO)(ENUM_TYPE_FROM enum_value_from, ENUM_TYPE_TO* enum_value_to) \
+    { \
+        switch (enum_value_from) \
+        { \
+            default: \
+                break; \
+            MU_FOR_EACH_2(CONVERT_ENUM_VALUE, __VA_ARGS__) \
+        } \
+        return MU_FAILURE; \
+    }
+
+// This macro simply adds the conversion of the "well known" by now _INVALID enum value
+#define MU_DEFINE_CONVERT_ENUM(ENUM_TYPE_FROM, ENUM_TYPE_TO, ...) \
+    MU_DEFINE_CONVERT_ENUM_WITHOUT_INVALID(ENUM_TYPE_FROM, ENUM_TYPE_TO, MU_C2(ENUM_TYPE_FROM, _INVALID), MU_C2(ENUM_TYPE_TO, _INVALID), __VA_ARGS__)
+
+#define CONSTRUCT_FROM_FAKE_ENUM(from_value, to_value) \
+    MU_C3(fake_from_, from_value, _C0FB3709_EDE8_4288_8F70_FDDB5D8D7A51),
+
+#define CONSTRUCT_TO_FAKE_ENUM(from_value, to_value) \
+    MU_C3(fake_to_, to_value, _C0FB3709_EDE8_4288_8F70_FDDB5D8D7A51),
+
+// This macro adds validation of the fact that the from values count in the list match the count of values in the
+// from enum type. It does that by having a special fake enum where all the "from" values are enumerated
+// The count check is done by having a zero sized array in a struct if the counts do not match
+#define MU_DEFINE_CONVERT_ENUM_WITH_VALIDATION(ENUM_TYPE_FROM, ENUM_TYPE_TO, ...) \
```

Comment:
>MU_DEFINE_CONVERT_ENUM_WITH_VALIDATION [](start = 8, length = 38)

What's the prevalence of scenarios where one doesn't want validation vs the scenarios where the user wants validation?

Poetry says this should be named MU_DEFINE_CONVERT_ENUM and the existing MU_DEFINE_CONVERT_ENUM should be named MU_DEFINE_CONVERT_ENUM_WITHOUT_VALIDATION.

Reply:
Sure, can do that.

---
In reply to: [394386388](https://github.com/Azure/azure-macro-utils-c/pull/36#discussion_r394386388) [](ancestors = 394386388)

Reply:
Done.

---
In reply to: [394619857](https://github.com/Azure/azure-macro-utils-c/pull/36#discussion_r394619857) [](ancestors = 394619857,394386388)